modMcf/src/org/aion/mcf/trie/TrieImpl.java

```diff
@@ -881,6 +882,64 @@ private void traceTrie(byte[] stateRoot, ScanAction action) {
     //        return true;
     //    }
 
+    @Override
+    public Map<ByteArrayWrapper, byte[]> getReferencedTrieNodes(byte[] value, int limit) {
+        CollectMappings collect = new CollectMappings();
+        ArrayList<byte[]> hashes = new ArrayList<>();
+        Value node = Value.fromRlpEncoded(value);
+
+        synchronized (cache) {
+            appendHashes(node, hashes);
+
+            while (!hashes.isEmpty() && collect.getSize() < limit) {
+                byte[] myHash = hashes.remove(0);
+                node = this.getCache().get(myHash);
+
+                if (node != null) {
+                    if (node.isList()) {
+                        List<Object> siblings = node.asList();
+                        if (siblings.size() == PAIR_SIZE) {
+                            Value val = new Value(siblings.get(1));
+                            if (val.isHashCode() && !hasTerminator((byte[]) siblings.get(0))) {
+                                hashes.add(val.asBytes());
+                            }
+                        } else {
+                            for (int j = 0; j < LIST_SIZE; ++j) {
+                                Value val = new Value(siblings.get(j));
+                                if (val.isHashCode()) {
+                                    hashes.add(val.asBytes());
+                                }
+                            }
+                        }
+                    }
+                    collect.doOnNode(myHash, node);
+                }
+            }
+        }
+        return collect.getNodes();
+    }
+
+    private void appendHashes(Value node, ArrayList<byte[]> hashes) {
+        if (node.isHashCode()) {
+            hashes.add(node.asBytes());
+        } else if (node.isList()) {
+            List<Object> siblings = node.asList();
+            if (siblings.size() == PAIR_SIZE) {
+                Value val = new Value(siblings.get(1));
+                if (val.isHashCode() && !hasTerminator((byte[]) siblings.get(0))) {
+                    hashes.add(val.asBytes());
+                }
+            } else {
+                for (int j = 0; j < LIST_SIZE; ++j) {
+                    Value val = new Value(siblings.get(j));
```

Comment:
Is it possible for siblings.size() to be neither PAIR_SIZE nor LIST_SIZE?

I'm just thinking, if siblings.size() is 3, then line 934 will get array out of bounds exception.

Reply:
It's not possible to have list nodes with different sizes other than the two above.